Wrong output for all_pairs_lowest_common_ancestor

I found a counterexample for the algorithm **all_pairs_lowest_common_ancestor** [(docs here)](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.lowest_common_ancestors.all_pairs_lowest_common_ancestor.html?highlight=lca).

The LCA of `(i,i)` must always be `i` regardless of the given DAG, but for the following DAG, it fails to meet this condition:

```python
import networkx as nx
from networkx.algorithms.lowest_common_ancestors import all_pairs_lowest_common_ancestor
print(nx.__version__)

G = nx.DiGraph()
G.add_nodes_from(range(5))
G.add_edges_from([(1, 0), (2, 0), (3, 2), (4, 1), (4, 3)])

for (i,j),a in all_pairs_lowest_common_ancestor(G):
    if i==j:
        assert a==i==j, f'Expected {i}. Got lca({i},{j}) = {a}'

```
Prints `2.5` and raises `AssertionError: Expected 0. Got lca(0,0) = 2`

You can check that the input is indeed a DAG manually, or like this:

```python
from networkx.algorithms.cycles import simple_cycles
assert list(simple_cycles(G))==[], 'Cycle found'
```

--- a/networkx/algorithms/tests/test_lowest_common_ancestors.py
+++ b/networkx/algorithms/tests/test_lowest_common_ancestors.py
@@ -316,6 +316,7 @@ class TestDAGLCA:
 
     def test_all_pairs_lca_one_pair_gh4942(self):
         G = nx.DiGraph()
+        # Note: order edge addition is critical to the test
         G.add_edge(0, 1)
         G.add_edge(2, 0)
         G.add_edge(2, 3)
@@ -323,3 +324,16 @@ class TestDAGLCA:
         G.add_edge(5, 2)
 
         assert nx.lowest_common_ancestor(G, 1, 3) == 2
+
+
+def test_all_pairs_lca_self_ancestors():
+    """Self-ancestors should always be the node itself, i.e. lca of (0, 0) is 0.
+    See gh-4458."""
+    # DAG for test - note order of node/edge addition is relevant
+    G = nx.DiGraph()
+    G.add_nodes_from(range(5))
+    G.add_edges_from([(1, 0), (2, 0), (3, 2), (4, 1), (4, 3)])
+
+    assert all(
+        u == v == a for (u, v), a in nx.all_pairs_lowest_common_ancestor(G) if u == v
+    )

